### PR TITLE
[Win32 Signals] Add term and ctrl-c signal handlers

### PR DESCRIPTION
**Pull request by davinci26**
Signed-off-by: Sotiris Nanopoulos <sonanopo@microsoft.com>

Part 1 of https://github.com/envoyproxy/envoy/issues/13188. Adds support for `ctrl+c` and `ctrl+break` for Envoy on Windows.

The implementation for this following:

1. On `platform_impl`  we register a `CtrlHandler` which runs on a separate thread.
2. On `signal_impl` we register a read event reader.

Thread (1) and (2) communicate via a socket pair and the event is handled on Windows the same way as it is handled on POSIX

Risk Level: Low (Windows Only)
Testing: Manual testing + integration tests
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by davinci26** [comment]
@envoyproxy/windows-dev

File: include/envoy/common/platform.h
Comment (wrowe) [inline include/envoy/common/platform.h:158]: It seems that CTRL_LOGOFF_EVENT and CTRL_SHUTDOWN_EVENT must follow the ENVOY_SIGTERM path as any linux would cause during shutdown, or envoy will have no chance to exit cleanly but os shutdown will be delayed.

The CTRL_BREAK_EVENT is really interesting; this could be mapped to graceful shutdown or hot restart.

**Review by wrowe — CHANGES_REQUESTED** [review]
This approach is otherwise looking good to me. In the shutdown/logoff cases, I've found that returning immediately from the CtrlHandler causes the typical "now attempt to kill" reaction immediately from the OS, so a delay is usually helpful.

File: test/exe/outofproc_main_test.py
Comment (sunjayBhatia) [inline test/exe/outofproc_main_test.py:70]: The assertion that we wait for a bit and the process has exited seems a little loose, if we were to remove the CTRL-BREAK handling logic wouldn't this test still pass?

Could we do some more assertions around maybe synchronizing the process exit or checking for some log line or stat or something that indicates graceful shutdown was invoked via signal? Some similar tests use a test main (instead of the real one) to do things like this, since we have log lines already in the dispatcher callback this might be doable without that?

File: include/envoy/common/platform.h
Comment (sunjayBhatia) [reply include/envoy/common/platform.h:158]: (Also CTRL SHUTDOWN is quite relevant for running Envoy via Windows containers as I believe that is the only notification/event a process will be able to handle when a container is being destroyed. Container orchestrators may also choose to do some lifecycle management with the Admin API `/quitquitquit` endpoint instead, but handling CTRL SHUTDOWN would cover other cases.)

File: test/exe/outofproc_main_test.py
Comment (wrowe) [reply test/exe/outofproc_main_test.py:70]: > The assertion that we wait for a bit and the process has exited seems a little loose

Agreed, we should determine our process has finished and then return, not some "pause 30 seconds" hack.

My understanding, ctrl+break/ctrl+close/ctrl+c do not have to be honored, so the timing of the return from those ctrl requests doesn't matter.

Shutdown/logout is time sensitive and manditory, so a process kill follows the return from the ctrl event. This behavior may have evolved since I last implemented it.

File: .bazelrc
Comment (sunjayBhatia) [inline .bazelrc:305]: Could you also remove this from `ci/windows_ci_steps.sh` just to clean this up

**Review by sunjayBhatia — CHANGES_REQUESTED** [review]

File: source/server/server.cc
Comment (sunjayBhatia) [inline source/server/server.cc:620]: Maybe out of scope of this PR and worth doing in a follow up, but it seems like this block could be abstracted out of this file at least, could enable some more detailed test coverage with mocking etc.

File: test/exe/outofproc_main_test.py
Comment (davinci26) [reply test/exe/outofproc_main_test.py:70]: > Some similar tests use a test main (instead of the real one) to do things like this, since we have log lines already in the dispatcher callback this might be doable without that?

I can parse the standard output to verify that the event was indeed caught and processed.

File: include/envoy/common/platform.h
Comment (antoniovicente) [reply include/envoy/common/platform.h:158]: Is your suggestion to register some additional signal events so they also go down the clean shutdown path?  Perhaps we should add a function that returns a list of signals that should trigger the clean shutdown handler

Should config for fast vs clean shutdown on CTRL_LOGOFF_EVENT and CTRL_BREAK_EVENT make sense?  I could see fast teardown being desirable on logoff.  Clean shutdown on both of these events may be most appropriate.

File: source/server/server.cc
Comment (antoniovicente) [inline source/server/server.cc:631]: See suggestion for platform functions that return a list of signals to associate with the clean shutdown handler as a possible alternative to ENVOY_SIGTERM and friends.

https://github.com/envoyproxy/envoy/pull/13954#discussion_r521584038

**Review by antoniovicente — COMMENTED** [review]
Thanks for the continued help to improve cross platform support.

File: source/server/server.cc
Comment (davinci26) [reply source/server/server.cc:631]: Agreed if the relationship is one-to-many then functions are the way to go. SIGTERM would map to multiple Windows events

**Comment by davinci26** [comment]
I am adding a /wait tag as I am working more with the automated test. @sunjayBhatia was right and the ctrl+break is caught from the test regardless of the signal handling. I am working on adding ctrl + c signals in the tests but it is a bit tricky.

**Comment by antoniovicente** [comment]
Let me know when it is ready for another round of review.

File: source/server/server.cc
Comment (davinci26) [inline source/server/server.cc:57]: unused

File: source/server/server.cc
Comment (davinci26) [inline source/server/server.cc:626]: revert POSIX changes

Comment (davinci26) [reply source/server/server.cc]: fixed in c484a7512620d167f4f04d6d13568e3714200ea5

File: source/server/server.cc
Comment (davinci26) [reply source/server/server.cc]: fixed in c484a7512620d167f4f04d6d13568e3714200ea5

File: test/exe/outofproc_main_test.py
Comment (davinci26) [reply test/exe/outofproc_main_test.py]: In terms of testing I did the following:

1. I added a "hacky" sleep statement that is responsible for waiting while Envoy is constructed. The alternative to that would be to keep pinging the Envoy ip/port and see when it responds back to know that it is fully initialized.
2. Wait for Envoy to close after we send the singal.
3. Read the standard output to verify that we closed with `ENVOY_WIN32_SIGTERM`

The test needs to be executed outside of bazel as a standalone executable because of the way the console signals work. Bazel is handling them for us and as a result the envoy process never sees them.

File: include/envoy/common/platform.h
Comment (davinci26) [reply include/envoy/common/platform.h]: All the signals in the control handler map to clean shutdown so I just created an event `ENVOY_WIN32_SIGTERM` and all the `CtrlHandler` events are mapped to it. 

All the Windows events are causing a clean shutdown so we dont need a function for now. When we add support for SCM if might have other events that are not related to shutdown and at that point we might need to implement a function

Comment (davinci26) [reply include/envoy/common/platform.h]: Is there a concept of clean vs fast shutdown?

File: .bazelrc
Comment (davinci26) [reply .bazelrc:305]: done

File: source/server/server.cc
Comment (sunjayBhatia) [inline source/server/server.cc:628]: nit: typo should be `SIGINT` (no change)

**Review by sunjayBhatia — CHANGES_REQUESTED** [review]

File: source/server/server.cc
Comment (sunjayBhatia) [reply source/server/server.cc:628]: (looks like that is a compile error anyway)

Comment (davinci26) [reply source/server/server.cc:628]: I saw that but I am on a different VM so I am waiting for a tiny bit to fix it.

Comment (davinci26) [reply source/server/server.cc]: fixed

**Comment by davinci26** [comment]
I expect the CI to go green now ( maybe with the exception of `clang-tidy`) @wrowe @antoniovicente ready for another review.

File: source/common/event/BUILD
Comment (antoniovicente) [inline source/common/event/BUILD:27]: This header is provided by :signal_lib, could we remove it from the dispatcher library?

File: source/common/event/BUILD
Comment (antoniovicente) [inline source/common/event/BUILD:64]: the strip_include_prefix may be a source of confusion, but there is prior art in source/common/common/BUILD and source/common/filesystem/BUILD so I guess this is fine.

I would have suggested having a signal_impl.h that includes the right platform header.

File: source/exe/BUILD
Comment (antoniovicente) [inline source/exe/BUILD:173]: Signal lib, not dispatcher lib

File: test/exe/BUILD
Comment (antoniovicente) [inline test/exe/BUILD:89]: I assume there's no explicit way to generate this test binary target only under windows.

File: include/envoy/common/platform.h
Comment (antoniovicente) [inline include/envoy/common/platform.h:68]: Could you update the issue title to something like:

win32 signal: add term and ctrl-c signal handlers

File: source/common/event/BUILD
Comment (davinci26) [reply source/common/event/BUILD:27]: I run into https://github.com/bazelbuild/bazel/issues/6337 if I remove it. This is why I added it

File: include/envoy/common/platform.h
Comment (davinci26) [reply include/envoy/common/platform.h:68]: You mean the PR title right? The corresponding issue is a bit more broad because we also want to enable SCM (service control manager) which is the second way to manage a process in Windows (by running it as a service)

File: test/exe/BUILD
Comment (davinci26) [reply test/exe/BUILD:89]: Based on bazel documentation 

>The manual tag keyword is treated specially. It marks the test_suite target as "manual" so that it will be ignored by the wildcard expansion and automated testing facilities.

So it should not have any negative side effect on the POSIX compilations units and it should be ignored. Also given that this test has the same deps as the other units in the compilation unit if you tried to compile the entire //test/exe build file it should be minimal overhead.

File: source/exe/BUILD
Comment (davinci26) [reply source/exe/BUILD:173]: thanks good catch! fixed

File: source/common/event/BUILD
Comment (antoniovicente) [reply source/common/event/BUILD:27]: Ah, ok.

File: include/envoy/common/platform.h
Comment (antoniovicente) [reply include/envoy/common/platform.h:68]: Yeah, sorry.  I meant PR title.

**Comment by sunjayBhatia** [comment]
I think you can just add to this: https://github.com/envoyproxy/envoy/blob/af17f954653afcab6a189673e3500b2e988f1aef/ci/run_clang_tidy.sh#L31-L35

There is a TODO to correctly run clang-tidy on Windows and MacOS

File: ci/run_clang_tidy.sh
Comment (sunjayBhatia) [inline ci/run_clang_tidy.sh:34]: I think you can leave out the leading `source/` on the addition to this line

Comment (davinci26) [reply ci/run_clang_tidy.sh]: thanks!

**Comment by davinci26** [comment]
@sunjayBhatia thanks for the suggestion! It worked.

**Comment by davinci26** [comment]
@wrowe and @antoniovicente PTAL when you get some time

File: source/common/event/win32/signal_impl.cc
Comment (antoniovicente) [inline source/common/event/win32/signal_impl.cc:36]: What guarantees that this insert into the flat_hash_map ends up being thread safe?

Consider implementing this as an array instead of a map since the number of possible signal numbers is small.

File: source/common/event/win32/signal_impl.cc
Comment (antoniovicente) [inline source/common/event/win32/signal_impl.cc:27]: possible future refactor: I wished that a socket pair wakeup mechanism existed as an primitive under source/common/event .  There's several scenarios where it could prove useful, including replacing the wakeup mechanism currently used for Dispatcher::post callbacks.

File: source/exe/win32/platform_impl.cc
Comment (antoniovicente) [inline source/exe/win32/platform_impl.cc:30]: A possible question: Why use a socket pair for wakeup instead of a higher level construct like a dispatcher post callback?  I think the answer is related to the set of actions that are allowed from a signal handler; writing to a socket is fine, taking locks is not and allocating memory is not fine.

Worth a comment?

File: source/common/event/win32/signal_impl.h
Comment (antoniovicente) [inline source/common/event/win32/signal_impl.h:33]: Is there a mechanism that clears ThreadSafeSingleton between tests?

File: test/exe/win32_outofproc_main_test.cc
Comment (antoniovicente) [inline test/exe/win32_outofproc_main_test.cc:67]: nit: What's the 1 in the test name about?

**Review by antoniovicente — COMMENTED** [review]
Change looks pretty much ready to merge.  Just a few minor questions.  Sorry for the delays in review, this one really went to the back burner.

File: source/common/event/win32/signal_impl.cc
Comment (davinci26) [reply source/common/event/win32/signal_impl.cc:27]: I assume the wake up mechanism is: `post_cb_->scheduleCallbackCurrentIteration();`. Makes sense, I would like to get #10871 first so we can have `AF_UNIX` socketpairs on windows and then I can add that.

I will open an issue for that so that we don't forget it in the meantime

File: source/exe/win32/platform_impl.cc
Comment (davinci26) [reply source/exe/win32/platform_impl.cc:30]: I added a comment.

 Initially my idea was that I did not want to have a global reference to the dispatcher. Seemed like a recipe for future disaster. Sadly this function does not allow for any additional context so communicating with a socket pair seems like the method that could cause the least amount of trouble. Also allocating memory or taking locks in this thread is risky.

File: source/common/event/win32/signal_impl.h
Comment (davinci26) [reply source/common/event/win32/signal_impl.h:33]: No it doesn't but I don't think it matters anyway. There are two reasons for that:

1. We only add new events. This means that the `map`/`array` is bounded in size.
2. There is no way to trigger these signals when you are running inside bazel or even gtest.

The only way to actually go through this code is by spawning envoy out of process. This is because the signals that we are catching are part of the console actually not signals to the process. And both bazel and gtest handle these console signals themselves so the code here is never executed.

File: source/common/event/win32/signal_impl.cc
Comment (antoniovicente) [reply source/common/event/win32/signal_impl.cc:27]: It is.  Schedulable callbacks are implemented on top of libevent timers which require taking a lock on the event loop in order to trigger activation.

Wakeup via a socket pair feels more elegant.

File: source/exe/win32/platform_impl.cc
Comment (antoniovicente) [reply source/exe/win32/platform_impl.cc:30]: Right.  At least the current implementation of Dispatcher post callbacks ends up taking locks on the libevent scheduler and the dispatcher.  Also, it allocates memory.

Comment (antoniovicente) [reply source/exe/win32/platform_impl.cc:30]: Also, I think you have some local changes that you have not pushed out yet.  I don't see the comments that you added.

File: test/exe/win32_outofproc_main_test.cc
Comment (davinci26) [reply test/exe/win32_outofproc_main_test.cc]: fixed in c378f5f6189624b029a4b5c17e677a25add84ce0

File: source/exe/win32/platform_impl.cc
Comment (davinci26) [reply source/exe/win32/platform_impl.cc]: Fixed now in c378f5f6189624b029a4b5c17e677a25add84ce0

I was compiling the whole thing last night and did not get a chance to have a full compilation and tests so I just answered the comments, and let the code change for the next day. Sorry for the confusion

File: source/common/event/win32/signal_impl.cc
Comment (davinci26) [reply source/common/event/win32/signal_impl.cc:32]: see #14244

File: source/common/event/win32/signal_impl.cc
Comment (davinci26) [reply source/common/event/win32/signal_impl.cc]: I assume the race condition that you refer to is someone hitting ctrl+c while the server main is adding new `signal_impl`. I transformed this into an array where each signal is assigned to a fixed position based on the signal id. Let me know what you think.

**Review by antoniovicente — APPROVED** [review]
Thanks!

**Comment by davinci26** [comment]
@envoyproxy/senior-maintainers can I also get a review from one of you!

File: source/common/event/win32/signal_impl.cc
Comment (ggreenway) [inline source/common/event/win32/signal_impl.cc:14]: Should this be a PANIC/RELEASE_ASSERT? Isn't this a strict programming error?

File: source/exe/win32/platform_impl.cc
Comment (ggreenway) [inline source/exe/win32/platform_impl.cc:14]: On posix this is supposed to be `sig_atomic_t` I think. Is that available on windows? `volatile` seems like the wrong qualifier here.

File: .bazelrc
Comment (ggreenway) [inline .bazelrc:301]: I don't object to this change, but it seems unrelated to the main topic of this PR. Can it be in a separate PR?

File: source/exe/win32/platform_impl.cc
Comment (ggreenway) [inline source/exe/win32/platform_impl.cc:27]: What do you mean by `Windows context`?

File: source/exe/win32/platform_impl.cc
Comment (ggreenway) [inline source/exe/win32/platform_impl.cc:33]: This allocates memory, contrary to the above comment. I think you can add an UnownedSlice to the buffer, and put the slice on the stack.

File: source/server/server.cc
Comment (ggreenway) [inline source/server/server.cc:648]: Can we make the SIGTERM handler for WIN32 and non-WIN32 the same? Maybe define `ENVOY_SIGTERM` on all platforms and use that?

**Review by ggreenway — CHANGES_REQUESTED** [review]

File: .bazelrc
Comment (davinci26) [reply .bazelrc:301]: The reason it is baked into this PR is because you can't build `//source/exe:envoy-static` without this definition so it is part of my inner dev loop.

No strong opinions though.

File: source/common/event/win32/signal_impl.cc
Comment (davinci26) [reply source/common/event/win32/signal_impl.cc:14]: My train of though here is the following:

Assume that this is actually a bug that is introduced and this `ENVOY_BUG` would hit. As a user of Envoy do I prefer to have reduced signal handling i.e. I can not close envoy from `ctrl+c` or do I want Envoy to fail to start. 

If I were the user I would just prefer the former because I can still work around it and have Envoy working in my server. I would have a bad experience for sure but I wouldn't be blocked.

Again no strong opinion and happy to change it to `PANIC` just stating were I am coming from.

File: source/exe/win32/platform_impl.cc
Comment (davinci26) [reply source/exe/win32/platform_impl.cc:14]: looking into it.

File: source/exe/win32/platform_impl.cc
Comment (davinci26) [reply source/exe/win32/platform_impl.cc:27]: This is one of the callbacks that are executed from a thread created and managed by the console host and not by Envoy. I bet `Windows context` is not clear enough, I will update it

File: source/exe/win32/platform_impl.cc
Comment (davinci26) [reply source/exe/win32/platform_impl.cc:33]: sure will do

File: source/server/server.cc
Comment (davinci26) [reply source/server/server.cc:648]: Will do that

File: source/exe/win32/platform_impl.cc
Comment (davinci26) [reply source/exe/win32/platform_impl.cc:33]: I dont see an overload of `add` that takes an `UnownedSlice` as input. There is `addBufferFragment` but that allocates the slice.

Should I just add an overload or a function `addUnownedSlice`?

Comment (davinci26) [reply source/exe/win32/platform_impl.cc:33]: Honestly I am not sure if this 1 byte allocation is going to save us if we are that close to OOM

File: source/exe/win32/platform_impl.cc
Comment (davinci26) [reply source/exe/win32/platform_impl.cc]: fixed by making it an atomic

File: source/server/server.cc
Comment (davinci26) [reply source/server/server.cc]: fixed

File: source/exe/win32/platform_impl.cc
Comment (davinci26) [reply source/exe/win32/platform_impl.cc]: fixed in latest commit.

**Comment by davinci26** [comment]
@ggreenway I addressed most of the comments and I am looking for some further clarification is some comments that I have left open. Thanks for the code review!

File: source/exe/win32/platform_impl.cc
Comment (antoniovicente) [reply source/exe/win32/platform_impl.cc:33]: I missed this issue.  The allocation of a buffer object ends up allocating memory too.  I would suggest use of writev instead of write from a buffer.

My understanding of the no-allocation rule in signal handlers is somewhat crude.  I think that the issue is that the allocation process itself takes locks; if the signal handler triggers while the thread is holding the allocator mutex the signal handler deadlocks.

File: source/common/event/win32/signal_impl.cc
Comment (antoniovicente) [reply source/common/event/win32/signal_impl.cc:14]: It does seem like a programming error.

Comment (davinci26) [reply source/common/event/win32/signal_impl.cc]: Fair. Changed it into a panic in this iteration.

File: source/exe/win32/platform_impl.cc
Comment (davinci26) [reply source/exe/win32/platform_impl.cc]: I changed it into `writev` and I opened another issue for some other windows threads that call `write` instead of `writev`

File: .bazelrc
Comment (ggreenway) [reply .bazelrc:301]: It's probably fine; I don't feel strongly about it.

**Review by ggreenway — APPROVED** [review]
LGTM

**Pull request merged**
